src/github.com/matrix-org/dendrite/mediaapi/fileutils/fileutils.go

```diff
@@ -0,0 +1,196 @@
+// Copyright 2017 Vector Creations Ltd
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fileutils
+
+import (
+	"bufio"
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/matrix-org/dendrite/mediaapi/types"
+)
+
+// GetPathFromBase64Hash evaluates the path to a media file from its Base64Hash
+// If the Base64Hash is long enough, we split it into pieces, creating up to 2 subdirectories
+// for more manageable browsing and use the remainder as the file name.
+// For example, if Base64Hash is 'qwerty', the path will be 'q/w/erty'.
+func GetPathFromBase64Hash(base64Hash types.Base64Hash, absBasePath types.Path) (string, error) {
+	var subPath, fileName string
+
+	hashLen := len(base64Hash)
+
+	switch {
+	case hashLen < 1:
+		return "", fmt.Errorf("Invalid filePath (Base64Hash too short): %q", base64Hash)
+	case hashLen > 255:
+		return "", fmt.Errorf("Invalid filePath (Base64Hash too long - max 255 characters): %q", base64Hash)
+	case hashLen < 2:
+		subPath = ""
+		fileName = string(base64Hash)
+	case hashLen < 3:
+		subPath = string(base64Hash[0:1])
+		fileName = string(base64Hash[1:])
```

Comment:
We could... this is kind of a relic from when I was writing the file based on a sanitised upload name. I can just check the length of the hash is >= 3 and error if not, in order to simplify the code. Done.